### PR TITLE
Implement Kademlia protocol

### LINKED ISSUE #211: As a Developer, I need to implement Kademlia protocol using internal libp2p implementation, so Rust node can discover other nodes.
Definition of DONE:
Rust node is able to, and can be used to, discover other nodes in a Mina network.

Now that #178 is almost ready for basic communications, we need to implement other protocols needed for participating in Mina network, namely Kademlia.

- [x] Plug-in to the stream handler for kademlia protocol
- [x] Sketch a test with a simple libp2p-based peer
- [x] Encode/decode Kademlia data
- [x] Implement routing table suitable for Kademlia
- [x] Handle remote requests
- [x] Implement Kademlia bootstrapping protocol
- [x] Support for multiple addresses
- [ ] Periodic bootstrap with random key
- [x] #380
- [x] Testing
- [ ] Big network simulation

Known issues:
- [x] #315
- [x] #361 
- [x] #382

### PR DESCRIPTION
**Pull request by akoptelov**
Fixes #211 

The most notable changes:
- Now `state.p2p.peers` are populated directly from the `config` instead of discovery/scheduler reducers.
- P2p-related effects (direct and time-based) are grouped to their own functions
- Yamux readiness condition is slightly modified (now it is reported as SelectDone for yamux is received)

I will add tests in a separate PR.

### CODE REVIEW

File: node/src/effects.rs [hunk 1]
```diff
@@ -40,68 +37,20 @@ pub fn effects<S: Service>(store: &mut Store<S>, action: ActionWithMeta) {
...
-                p2p_try_reconnect_disconnected_peers(store);
-                p2p_request_best_tip_if_needed(store);
-                p2p_request_snarks_if_needed(store);
+            p2p_timeout_effects(store, &meta);
 
-                store.dispatch(P2pDiscoveryAction::KademliaBootstrap);
-                store.dispatch(P2pDiscoveryAction::KademliaInit);
-            }
+            p2p_request_best_tip_if_needed(store);
+            // p2p_request_snarks_if_needed(store);
```
Comment (tizoc) [inline node/src/effects.rs:43]: @akoptelov is this one not working anymore?

Comment (akoptelov) [reply node/src/effects.rs:43]: Let me check

**Comment by tizoc** [comment]
@akoptelov unrelated to your changes, but there are two files that have a typo in their filename (`p2p_networl_select*` instead of `p2p_network_select`)

File: p2p/src/network/scheduler/p2p_network_scheduler_reducer.rs [hunk 2]
```diff
@@ -86,15 +89,15 @@ impl P2pNetworkSchedulerState {
...
+                                ..Default::default()
+                            }));
                     }
                     Some(token::Protocol::Stream(stream_kind)) => {
-                        let Some(stream_id) = a.kind.stream_id() else {
-                            return;
-                        };
-                        let Some(peer_id) = a.kind.peer_id() else {
+                        let SelectKind::Stream(peer_id, stream_id) = a.kind else {
+                            error!(meta.time(); "incorrect stream kind for protocol stream: {stream_kind:?}");
```
Comment (tizoc) [inline p2p/src/network/scheduler/p2p_network_scheduler_reducer.rs:100]: This should be standard for all our early returns inside reducers

File: p2p/src/network/kad/p2p_network_kad_internals.rs [hunk 3]
```diff
@@ -0,0 +1,720 @@
...
+}
+
+impl P2pNetworkKadEntry {
+    pub fn new(peer_id: PeerId, addrs: Vec<Multiaddr>) -> Self {
+        let key = peer_id.clone().into();
+        P2pNetworkKadEntry {
+            key,
+            peer_id,
+            addrs,
+            // TODO
```
Comment (tizoc) [inline p2p/src/network/kad/p2p_network_kad_internals.rs:351]: What is this TODO about? also, is `Connected` a good default?

File: p2p/src/network/kad/stream/p2p_network_kad_stream_effects.rs [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+                | A::SendReply {
+                    addr,
+                    peer_id,
+                    stream_id,
+                    ..
+                },
+                S::OutgoingBytes { bytes, .. },
+            ) => {
+                // send data to the network
+                store.dispatch(P2pNetworkYamuxOutgoingDataAction {
```
Comment (tizoc) [inline p2p/src/network/kad/stream/p2p_network_kad_stream_effects.rs:99]: Unrelated to this PR, but this is still using old-style actions.

**Review by tizoc — APPROVED** [review]
@akoptelov I don't have the required background to give meaningful feedback, but left a few comments (not necessarily actionable stuff). In general it looks good to me.

**Comment by tizoc** [comment]
Btw, rebasing my work on top of this to give it a test

File: p2p/src/network/kad/p2p_network_kad_effects.rs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                _,
+            ) => {
+                let kad_key = key.into();
+                let closer_peers = state
+                    .routing_table
+                    .find_node(&kad_key)
+                    .cloned()
+                    .collect::<Vec<_>>();
+
+                debug!(meta.time(); "found {} peers", closer_peers.len());
```
Comment (tizoc) [inline p2p/src/network/kad/p2p_network_kad_effects.rs:54]: This `debug` is not imported, makes compilation fail.

**Review by tizoc — CHANGES_REQUESTED** [review]
Found a small issue, missing import.

File: p2p/src/network/kad/bootstrap/p2p_network_kad_bootstrap_actions.rs [hunk 6]
```diff
@@ -0,0 +1,47 @@
+use redux::EnablingCondition;
+use serde::{Deserialize, Serialize};
+
+use crate::{P2pAction, P2pNetworkKadAction, P2pNetworkKadLatestRequestPeers, P2pState, PeerId};
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum P2pNetworkKadBootstrapAction {
+    CreateRequests {},
```
[the remark this answers is not in the document]
Comment (akoptelov) [reply p2p/src/network/kad/bootstrap/p2p_network_kad_bootstrap_actions.rs:8]: Seems like there were parameters, but now they're gone...

**Review by tizoc — APPROVED** [review]
LGTM

**Pull request merged**